lib/pbench/cli/agent/commands/results/move.py

```diff
@@ -0,0 +1,257 @@
+import os
+import shutil
+import socket
+import tempfile
+
+import click
+
+from pbench.agent.base import BaseCommand
+from pbench.agent.results import MakeResultTb, CopyResultTb
+from pbench.cli.agent import pass_cli_context, CliContext
+from pbench.cli.agent.options import common_options
+from pbench.common.exceptions import BadMDLogFormat
+from pbench.common.utils import validate_hostname
+
+
+class MoveResults(BaseCommand):
+    """MoveResutls - command implementation for "pbench results move."
+
+    This command replaces the previous (legacy) interface,
+    `pbench-move-results`, and its sibling, `pbench-copy-results`, with no
+    attempt to implement any backwards compatibility with that legacy
+    interface.
+
+    This command is responsible for finding all the existing pbench data
+    collection on the local host (controller), package them up as a tar ball,
+    and send to the remote pbench server.
+    """
+
+    def __init__(self, context: CliContext):
+        super().__init__(context)
+
+    def execute(self, single_threaded: bool, copy_only: bool = False) -> int:
+        runs_copied = 0
+        failures = 0
+        no_of_tb = 0
+
+        with tempfile.TemporaryDirectory(
+            dir=self.config.pbench_tmp, prefix="pbench-move-results."
+        ) as temp_dir:
+            for dirent in self.config.pbench_run.iterdir():
+                if not dirent.is_dir():
+                    continue
+                if dirent.name.startswith("tools-") or dirent.name == "tmp":
+                    continue
+
+                no_of_tb += 1
+                result_dir = dirent
+
+                try:
+                    mrt = MakeResultTb(
+                        result_dir,
+                        temp_dir,
+                        self.context.controller,
+                        self.config,
+                        self.logger,
+                    )
+                except (
+                    NotADirectoryError,
+                    FileNotFoundError,
+                    MakeResultTb.AlreadyCopied,
+                    MakeResultTb.BenchmarkRunning,
+                ) as exc:
+                    self.logger.error(str(exc))
+                    failures += 1
+                    continue
```

Comment:
Not sure we can improve this much.  If an exception does blow it out of the water (the ValueError won't because we validate the host name before calling `execute()`) then the catch all exception around `execute()` will handle it.

Reply:
Having the outer-most catch-all exception handler catch the error is fine (I think...), provided it can report enough information and context for us to diagnose the issue.  (Yeah, the `ValueError` is an interesting case, because we _want_ it to blow all the way back to the user, but, as you say, we explicitly forestall that by checking the controller value up front, so we don't really need to consider it here.)

We want to catch all the recoverable or result-specific failures here while letting the unrecoverable (and unexpected) ones blow by, and the question is, how do we tell the difference.  We can catch an explicit list here, but I think that's fragile; we can catch `Exception`, but that will catch the unrecoverable/unexpected ones which may or may not be OK; or, we can try to have `MakeResultTb` produce the list of exceptions to catch (presumably via a base class, but it could just export a tuple), which would do the job without violating encapsulation, and it would put the burden of maintaining the list in the appropriate place (but, it's still yucky).